kratos/utilities/math_utils.h

```diff
@@ -699,29 +713,66 @@ class KRATOS_API(KRATOS_CORE) MathUtils
     }
 
     /**
-     * @brief Clamps a value between a minimum and maximum range.
-     * @details This function ensures that the given value `rX` lies within the specified
-     * range `[rMinimum, rMaximum]`. If `rX` is less than `rMinimum`, it returns
-     * `rMinimum`. If `rX` is greater than `rMaximum`, it returns `rMaximum`.
-     * Otherwise, it returns `rX`.
-     * @tparam T The type of the value and bounds, typically a numeric type.
-     * @param rX The value to clamp.
-     * @param rMinimum The minimum bound.
-     * @param rMaximum The maximum bound.
-     * @return The clamped value.
-     */
+    * @brief Clamps a value between a minimum and maximum range.
+    * @details This function ensures that the given value `rX` lies within the specified
+    * range `[rMinimum, rMaximum]`. If `rX` is less than `rMinimum`, it returns
+    * `rMinimum`. If `rX` is greater than `rMaximum`, it returns `rMaximum`.
+    * Otherwise, it returns `rX`.
+    * @tparam T The type of the value and bounds, typically a numeric type.
+    * @param rX The value to clamp.
+    * @param rMinimum The minimum bound.
+    * @param rMaximum The maximum bound.
+    * @param rTolerance The tolerance value to consider for clamping, defaults to epsilon.
+    * @return The clamped value.
+    */
+    template <typename T>
+    static inline T Clamp(
+        const T& rX,
+        const T& rMinimum,
+        const T& rMaximum,
+        const T& rTolerance = ZeroTolerance
+        )
+    {
+        if (rX < rMinimum - rTolerance) {
+            return rMinimum;
+        } else if (rX > rMaximum + rTolerance) {
+            return rMaximum;
+        } else {
+            return rX;
+        }
+    }
+
+    /**
+    * @brief Clamps a value between a minimum and maximum range with a scenario.
+    * @details This function ensures that the given value `rX` lies within the specified
+    * range `[rMinimum, rMaximum]`. If `rX` is less than `rMinimum`, it returns
+    * `rMinimum`. If `rX` is greater than `rMaximum`, it returns `rMaximum`.
+    * Otherwise, it returns `rX`. It also returns a scenario indicating if the value is below the minimum, above the maximum or within the bounds.
+    * @tparam T The type of the value and bounds, typically a numeric type.
+    * @param rX The value to clamp.
+    * @param rMinimum The minimum bound.
+    * @param rMaximum The maximum bound.
+    * @param rScenario The scenario where the value is clamped. It can be BELOW_MINIMUM, ABOVE_MAXIMUM or WITHIN_BOUNDS.
+    * @param rTolerance The tolerance value to consider for clamping, defaults to epsilon.
+    * @return The clamped value.
+    */
     template <typename T>
     static inline T Clamp(
         const T& rX,
         const T& rMinimum,
-        const T& rMaximum
+        const T& rMaximum,
+        ClampScenario& rScenario,
+        const T& rTolerance = ZeroTolerance
```

Comment:
tolerance kinda does not make sense here

then the order becomes important, if you check first upper then lower or vise-versa
I would remove it

Reply:
I was thinking that it was easier to just move the minium or maximum by tolerance

Reply:
I am not sure if I understand correctly the @philbucher comment. The tolerance expands the span which can be very useful in geometrical operations. I am not sure how this can be affected by order of the check.